File: docs/sphinx/snippets/python/using/examples/two_qubit_custom_op.py
# ============================================================================ #
# Copyright (c) 2022 - 2025 NVIDIA Corporation & Affiliates.                   #
# All rights reserved.                                                         #
#                                                                              #
# This source code and the accompanying materials are made available under     #
# the terms of the Apache License 2.0 which accompanies this distribution.     #
# ============================================================================ #

# `fmt: off`
#[Begin Docs]
import cudaq
import numpy as np

# Create and test a custom CNOT operation.
cudaq.register_operation("my_cnot", np.array([1, 0, 0, 0,
                                              0, 1, 0, 0,
                                              0, 0, 0, 1,
                                              0, 0, 1, 0]))

@cudaq.kernel
def bell_pair():
    qubits = cudaq.qvector(2)
    h(qubits[0])
    my_cnot(qubits[0], qubits[1]) # `my_cnot(control, target)`


cudaq.sample(bell_pair).dump() # prints { 11:500 00:500 } (exact numbers will be random)


# Construct a custom unitary matrix for X on the first qubit and Y
# on the second qubit.
X = np.array([[0,  1 ], [1 , 0]])
Y = np.array([[0, -1j], [1j, 0]])
XY = np.kron(X, Y)

# Register the custom operation
cudaq.register_operation("my_XY", XY)

@cudaq.kernel
def custom_xy_test():
    qubits = cudaq.qvector(2)
    my_XY(qubits[0], qubits[1])
    y(qubits[1]) # undo the prior Y gate on qubit 1


cudaq.sample(custom_xy_test).dump() # prints { 10:1000 }
#[End Docs]
# `fmt: on`

File: pr-4962/sphinx/snippets/cpp/using/two_qubit_custom_op.cpp
/*******************************************************************************
 * Copyright (c) 2022 - 2026 NVIDIA Corporation & Affiliates.                  *
 * All rights reserved.                                                        *
 *                                                                             *
 * This source code and the accompanying materials are made available under    *
 * the terms of the Apache License 2.0 which accompanies this distribution.    *
 ******************************************************************************/

// [Begin Docs]
#include <cudaq.h>

CUDAQ_REGISTER_OPERATION(MyCNOT, 2, 0,
                         {1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0});

CUDAQ_REGISTER_OPERATION(
    MyXY, 2, 0,
    {0, 0, 0, {0, -1}, 0, 0, {0, 1}, 0, 0, {0, -1}, 0, 0, {0, 1}, 0, 0, 0});

__qpu__ void bell_pair() {
  cudaq::qubit q, r;
  h(q);
  MyCNOT(q, r); // MyCNOT(control, target)
}

__qpu__ void custom_xy_test() {
  cudaq::qubit q, r;
  MyXY(q, r);
  y(r); // undo the prior Y gate on qubit 1
}

int main() {
  auto counts = cudaq::sample(bell_pair);
  counts.dump(); // prints { 11:500 00:500 } (exact numbers will be random)

  counts = cudaq::sample(custom_xy_test);
  counts.dump(); // prints { 10:1000 }
}
// [End Docs]
